From 47ed0c7cdcb4d516e41769e38a9cd5ef4d01713d Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 25 May 2014 14:57:03 +0000
Subject: [PATCH] [BEANUTILS-458] Added a check for null result objects.

A null input object should be accepted by a converted without throwing a
ConversionException. Thanks to Manuel Dominguez Sarmiento for the proposed fix.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/beanutils/trunk@1597432 13f79535-47bb-0310-9956-ffa450edef68
---
 .../beanutils/locale/BaseLocaleConverter.java |  4 ++
 .../beanutils/bugs/Jira458TestCase.java       | 57 +++++++++++++++++++
 2 files changed, 61 insertions(+)
 create mode 100644 src/test/java/org/apache/commons/beanutils/bugs/Jira458TestCase.java

diff --git a/src/main/java/org/apache/commons/beanutils/locale/BaseLocaleConverter.java b/src/main/java/org/apache/commons/beanutils/locale/BaseLocaleConverter.java
index f650efde0..f34a74eab 100644
--- a/src/main/java/org/apache/commons/beanutils/locale/BaseLocaleConverter.java
+++ b/src/main/java/org/apache/commons/beanutils/locale/BaseLocaleConverter.java
@@ -281,6 +281,10 @@ private static <T> T checkConversionResult(Class<T> type, Object result) {
             T temp = (T) result;
             return temp;
         }
+
+        if (result == null) {
+            return null;
+        }
         if (type.isInstance(result)) {
             return type.cast(result);
         }
diff --git a/src/test/java/org/apache/commons/beanutils/bugs/Jira458TestCase.java b/src/test/java/org/apache/commons/beanutils/bugs/Jira458TestCase.java
new file mode 100644
index 000000000..b8f1646cf
--- /dev/null
+++ b/src/test/java/org/apache/commons/beanutils/bugs/Jira458TestCase.java
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.beanutils.bugs;
+
+import java.util.Locale;
+
+import junit.framework.TestCase;
+
+import org.apache.commons.beanutils.Converter;
+import org.apache.commons.beanutils.locale.converters.IntegerLocaleConverter;
+
+/**
+ * BaseLocaleConverter.checkConversionResult() fails with ConversionException when result
+ * is null when it should not.
+ *
+ * @version $Id$
+ * @see <a href="https://issues.apache.org/jira/browse/BEANUTILS-458">https://issues.apache.org/jira/browse/BEANUTILS-458</a>
+ */
+public class Jira458TestCase extends TestCase {
+    /**
+     * Helper method for testing a conversion with null as default.
+     *
+     * @param input the input string
+     */
+    private void checkConversionWithNullDefault(String input) {
+        Converter converter = new IntegerLocaleConverter(null, Locale.US);
+        assertNull("Wrong result", converter.convert(Integer.class, input));
+    }
+
+    /**
+     * Tests a conversion passing in null.
+     */
+    public void testConversionWithNullDefaultNullInput() {
+        checkConversionWithNullDefault(null);
+    }
+
+    /**
+     * Tests a conversion passing in an empty string.
+     */
+    public void testConversionWithNullDefaultEmptyString() {
+        checkConversionWithNullDefault("");
+    }
+}
